core/pkg/monitor/openmetrics.go

```diff
@@ -0,0 +1,351 @@
+package monitor
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/wandb/wandb/core/internal/clients"
+	"github.com/wandb/wandb/core/pkg/observability"
+	spb "github.com/wandb/wandb/core/pkg/service_go_proto"
+
+	"github.com/prometheus/common/expfmt"
+)
+
+// Filter represents a filter to apply to metrics collected from an OpenMetrics endpoint.
+//
+// Provides a fine-grained control over which metrics to collect.
+// OpenMetrics filters can be specified via wandb.Settings in one of the two formats:
+//   - {"metric regex pattern, including endpoint name as prefix": {"label": "label value regex pattern"}}
+//   - ("metric regex pattern 1", "metric regex pattern 2", ...)
+//
+// These are converted to a list of Filter objects using the processMappingFilters
+// and processSequenceFilters functions, respectively.
+//
+// The following code snippet demonstrates how to configure OpenMetrics filters
+// to capture only the metrics DCGM_FI_DEV_POWER_USAGE with the labels gpu="0" or gpu="1"
+// from node1 and 5 metrics with the label gpu=".*" from node2.
+//
+// run = wandb.init(
+//
+//	project="dcgm",
+//	settings=wandb.Settings(
+//	    _stats_open_metrics_endpoints={
+//	        "node1": "http://192.168.0.1:9400/metrics",
+//	        "node2": "http://192.168.0.2:9400/metrics",
+//	    },
+//	    _stats_open_metrics_filters={
+//	        "node1.DCGM_FI_DEV_POWER_USAGE": {
+//	            "gpu": "[0,1]",
+//	        },
+//	        "node2.DCGM_FI_DEV_(POWER_USAGE|MEM_COPY_UTIL|TOTAL_ENERGY_CONSUMPTION|GPU_TEMP|MEMORY_TEMP)": {
+//	            "gpu": ".*",
+//	        },
+//	    },
+//	),
+//
+// )
```

Comment:
maybe just describe the format of the input, i think this is too tied to what happens on the client side?